src/Symfony/Component/HttpClient/RetryableHttpClient.php

```diff
@@ -54,40 +55,72 @@ public function __construct(HttpClientInterface $client, RetryDeciderInterface $
+        return new AsyncResponse($this->client, $method, $url, $options, function (ChunkInterface $chunk, AsyncContext $context) use ($method, $url, $options, &$retryCount, &$content, &$firstChunk) {
             $exception = null;
             try {
                 if ($chunk->isTimeout() || null !== $chunk->getInformationalStatus()) {
                     yield $chunk;
 
                     return;
                 }
-
-                // only retry first chunk
-                if (!$chunk->isFirst()) {
-                    $context->passthru();
-                    yield $chunk;
-
-                    return;
-                }
             } catch (TransportExceptionInterface $exception) {
                 // catch TransportExceptionInterface to send it to strategy.
             }
 
-            $statusCode = $context->getStatusCode();
-            $headers = $context->getHeaders();
-            if ($retryCount >= $this->maxRetries || !$this->decider->shouldRetry($method, $url, $options, $partialResponse = new MockResponse($context->getContent(), ['http_code' => $statusCode, 'headers' => $headers]), $exception)) {
+            if ($retryCount >= $this->maxRetries) {
                 $context->passthru();
                 yield $chunk;
 
                 return;
             }
 
+            $statusCode = $context->getStatusCode();
+            $headers = $context->getHeaders();
+            if ($chunk instanceof FirstChunk || null !== $exception) {
+                $shouldRetry = $this->decider->shouldRetry($method, $url, $options, $context->getStatusCode(), $context->getHeaders(), null, $exception);
+
+                if (false === $shouldRetry) {
+                    $context->passthru();
+                    yield $chunk;
+
+                    return;
+                }
+
+                // Decider need body to decide
+                if (null === $shouldRetry) {
+                    $firstChunk = $chunk;
+                    $content = '';
+                    // disable exception thrown on destruct
+                    $context->getResponse()->getStatusCode();
+
+                    return;
+                }
+            } else {
+                $content .= $chunk->getContent();
+                if (!$chunk instanceof LastChunk) {
+                    return;
+                }
+                $shouldRetry = $this->decider->shouldRetry($method, $url, $options, $context->getStatusCode(), $context->getHeaders(), $content, $exception);
+                if (null === $shouldRetry) {
+                    throw new \LogicException(sprintf('The method "shouldRetry" of "%s" must not return null when called with a body.', \get_class($this->decider)));
+                }
+
+                if (false === $shouldRetry) {
+                    $context->passthru();
+                    yield $firstChunk;
+                    yield $context->createChunk($content);
+
```

Comment:
missing `$content = '';`

Reply:
I don't think it's necessary. 
`$context->passthru();` (called 2 lines above) will reset passthru and this closure will never be called again. The `$content` variable should not be used anymore, and should be destruct.

Reply:
still better explicit :)